### ISSUE TITLE
Feed jscpd All Files At Once

### ISSUE THREAD

**Issue by Kurt-von-Laven** [github/super-linter#1619]
## Describe the bug

jscpd's purpose is the detection of duplicated code, so invoking it separately on each individual file defeats its purpose by limiting it to only detecting duplicate code within each individual file.

## Expected behavior

jscpd is called one time on the entire code base, so it catches duplication between files, and it's threshold setting works properly. It doesn't make much sense to run it only on changed files either in my opinion, because then it will miss all the cases where code was copied from an unmodified file to a file in the pull request. Perhaps the ideal thing would be to only report duplication involving the modified files, but that seems like way too much work for one issue.

## Steps to Reproduce

1. Create a .github/linters/.jscpd.json.
2. Create some duplicate files that lack duplication within them.
3. Open a pull request, or take whatever alternative action is needed to trigger the action.
4. No error is reported, regardless of whether the entire code base is linted or only the modified files.

**Comment by admiralAwkbar** [comment]
@Kurt-von-Laven 
This is an interesting point...

Usually, we check the single file for duplicate code which is a decent practice, but I can understand the idea of checking all the code for duplicates.

My guess on checking all code in the codebase would lead to a lot of markdown, text, and template files kicking off false positives in this case, but I would assume most code should have too many duplicates

@ferrarimarco  @GaboFDC  any thoughts on this?

**Comment by Kurt-von-Laven** [comment]
@ferrarimarco, I see the rationale for re-classifying this as a feature request. Yes, that does seem like the best path forward to me as well.

@admiralAwkbar, it is certainly possible that jscpd will have undesirable behavior out of the box on some projects, but we didn't encounter many issues with running it on our repositories after a small amount of configuration. The tool itself is highly configurable. We are just one project, but most of duplication I was able to uncover using this tool was between files rather than within files. Presently, we just tell Super-Linter not to run it and then run it ourselves.

For the benefit of future readers of this discussion, I will pause to say that copy/paste detection within a small file is a trivial problem compared to the one jscpd addresses with its sophisticated [Rabin-Karp](https://en.wikipedia.org/wiki/Rabin%E2%80%93Karp_algorithm) implementation. It is clearly the intent of jscpd to be run on all file at once, or they would not have gone to such great lengths.

Your point about false positives got me thinking about the [default configuration file](https://github.com/github/super-linter/blob/master/TEMPLATES/.jscpd.json). I might add `"console"` after `"consoleFull"` (order matters) so it produces the nice table summarizing the results at the end, use `gitignore: true` so it ignores what you .gitignore, and possibly add super-linter.log to the ignore list. Happy to open a PR if those sound like desirable changes.

**Comment by ferrarimarco** [comment]
@Kurt-von-Laven I think you've a point here actually.

@admiralAwkbar are there other linters that would benefit from this? IIRC not, but maybe I'm wrong. We have so many now :)

**Comment by admiralAwkbar** [comment]
@ferrarimarco  i think this is `go`s major issue...

So at least those 2 would be a win...

I would say that making a flag that is `<LANGUAGE>_LINTER_SINGLE_RUN` or something similar would help solve for those cases.... could be a somewhat simple fix... or it could be a nightmare lol

**Comment by ferrarimarco** [comment]
Having that kind of variable (and sensible defaults) might help avoiding breaking changes. The (good) thing is that I expect this change to be mostly limited to the "build file list" logic, and maybe minor tweaks to linters invocation.

**Comment by iesoftwaredeveloper** [comment]
Just to add another side affect of invoking jscpd on each file.  If you are creating an html report using the `"reporters": ["html"]` in the `.jscpd.json` file then the report for only the most recent file to be checked is saved.  All other reports are overwritten by the most recent to be generated.  This has the same effect for other reporters as well such as markdown, xml and json.

This results in an in ability to use the reporters feature to resolve duplicates.

A possible fix to this would be to pass the name of the output report based upon the file being checked.  However, this would result in possibly hundreds of files instead of a single report.  Perhaps an option to run on the entire code base would allow the current functionality to be used while allowing for a single call when desired.

**Comment by ferrarimarco** [comment]
Right, another reason to implement this

**Comment by Kurt-von-Laven** [comment]
Should I go ahead with this (very small) proposal?

> Your point about false positives got me thinking about the [default configuration file](https://github.com/github/super-linter/blob/master/TEMPLATES/.jscpd.json). I might add `"console"` after `"consoleFull"` (order matters) so it produces the nice table summarizing the results at the end, use `gitignore: true` so it ignores what you .gitignore, and possibly add super-linter.log to the ignore list. Happy to open a PR if those sound like desirable changes.

**Comment by yxliang01** [comment]
Feeding all supported files to linters is a much needed feature. Lots of supported linters support this and it should improve a lot the performance when number of files is large.

**Comment by nvuillam** [comment]
@yxliang01 if you are in a hurry, it's already implemented on Super-Linter little brother [Mega-Linter](https://nvuillam.github.io/mega-linter/) :) (all files in an unique jscpd call + all files in an unique linter call for most of linters )

**Comment by Kurt-von-Laven** [comment]
*shoo stale bot. shoo.*

**Comment by Kurt-von-Laven** [comment]
I realize this is an optimistic proposal, but I wonder whether Super-Linter would have any interest in joining forces with [Mega-Linter](https://megalinter.github.io/) in some way? I don't have a pre-conceived notion of how such a collaboration would work, but I am hoping to initiate a dialogue (and happy to move this discussion to a more appropriate forum as desired).

Here are some of [the major highlights](https://megalinter.github.io/mega-linter-vs-super-linter/): 
- Ported from Bash to Python.
- Extends the slim image concept to [a wide variety of reusable flavors](https://megalinter.github.io/flavors/) to minimize download time.
- Runs linters in parallel for faster performance.
- Batches files for linters that support it for faster performance (and correctness in the case of jscpd as discussed above).
- Can automatically apply the fixes of those linters that support it.
- Offers [a local runner](https://megalinter.github.io/mega-linter-runner/) (made feasible by flavors) if you don't want to wait in a CI queue or desire to run linters as a [pre-commit](https://pre-commit.com/) hook.
- Supports multiple report formats (automatically stored as artifacts by [the accompanying GitHub Action](https://github.com/marketplace/actions/mega-linter)).
- Offers a CLI installation and configuration assistant.
- Has enhanced documentation.
- Generates recommended IDE extensions.
- Supports plugins for less widely used linters.
- Features a simplified architecture to reduce maintenance burden and increase automation.
- Has [87% test coverage](https://app.codecov.io/gh/megalinter/megalinter).

That being said Super-Linter has GitHub's brand recognition and backing, which means greater visibility, more users, more contributors, more maintainer FTEs, more stars, etc. It is worth noting that Mega-Linter is licensed AGPL-3.0, while Super-Linter is licensed MIT, so some legal consideration would need to be put into any effort to collaborate or unite the time. Since Mega-Linter began as a hard fork of Super-Linter (#791), there is a great deal of overlap between the two's supported linters. I have summarized the differences below.

Super-Linter (but not Mega-Linter) includes:
- [clang-format](https://clang.llvm.org/docs/ClangFormatStyleOptions.html)
- [eslint-plugin-json](https://github.com/azeemba/eslint-plugin-json#readme)
- [gitleaks (credentials)](https://github.com/zricethezav/gitleaks#readme)
- [google-java-format](https://github.com/google/google-java-format#google-java-format)
- [rustfmt](https://github.com/rust-lang/rustfmt#readme)
- [textlint (natural language)](https://textlint.github.io/)

Mega-Linter (but not Super-Linter) includes:
- [bandit (Python security)](https://bandit.readthedocs.io/en/latest/)
- [checkov (Terraform, CloudFormation, Kubernetes, Helm, ARM Templates, and Serverless)](https://www.checkov.io/)
- [cspell (spelling)](https://cspell.org/)
- [git-diff (merge conflict markers)](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---check)
- [graphql-schema-linter](https://github.com/cjoudrey/graphql-schema-linter#readme)
- [jsonlint](https://github.com/zaach/jsonlint#readme)
- [markdown-link-check](https://github.com/tcort/markdown-link-check#readme)
- [markdown-table-formatter](https://github.com/nvuillam/markdown-table-formatter#readme)
- [misspell (spelling)](https://github.com/client9/misspell#readme)
- [mypy (Python typing)](https://mypy.readthedocs.io/en/stable/)
- [puppet-lint](http://puppet-lint.com/)
- [remark-lint](https://github.com/remarkjs/remark-lint#readme)
- [restructuredtext-lint](https://github.com/twolfson/restructuredtext-lint#readme)
- [revive (Go)](https://revive.run/)
- [rstcheck (reStructuredText syntax)](https://github.com/myint/rstcheck#readme)
- [rstfmt (reStructuredText formatter)](https://github.com/dzhu/rstfmt)
- [sfdx-scanner (Apex)](https://forcedotcom.github.io/sfdx-scanner/)
- [scalafix](https://scalacenter.github.io/scalafix/)
- [scss-lint](https://github.com/sds/scss-lint#readme)
- [Secretlint (credentials)](https://github.com/secretlint/secretlint#readme)
- [swiftlint](https://github.com/realm/SwiftLint#readme)
- [terraform-fmt](https://www.terraform.io/docs/cli/commands/fmt.html)
- [tsqllint](https://github.com/tsqllint/tsqllint#readme)
- [v8r (JSON and YAML)](https://github.com/chris48s/v8r#readme)

There is so much overlap between these two projects that I anticipate a great deal of duplicated effort between our two maintenance teams. 3 weeks after you added GitLeaks [we were asked to add it as well GitLeaks](https://github.com/megalinter/megalinter/issues/876). Conversely, I asked for batching support above before I learned of Mega-Linter. I expect our communities to often request the same thing from one of us that the other just implemented. I submit that neither one of our projects alone will ever be as good as both together, and our joint communities would be best served by our joint efforts. Disclosure: I didn't write Mega-Linter (the credit belongs to @nvuillam and many contributors), but I am a happy user that recently become the first co-maintainer.

**Comment by admiralAwkbar** [comment]
@Kurt-von-Laven Love the idea.

Not sure how to implement that but would love to come together on this.

The reason for the hard split was the move from bash to python. At that point we wanted to keep the code as simple as possible. 

I would completely agree, we are both doing the exact same thing in the wild, and somehow coming together would bring a lot to the open-source community.

@zkoppert @nemchik @Hanse00 @GaboFDC @ferrarimarco

What are yall's thoughts on this idea?

**Comment by Kurt-von-Laven** [comment]
Thanks, @admiralAwkbar :). It was a pleasure working with you back when I was a Super-Linter user. In retrospect, this discussion deserves its own issue, so moving this to #2090 to avoid cannibalizing this issue.
